lib/go-tc/users.go

```diff
@@ -101,6 +110,179 @@ type UserCurrent struct {
+	GID                json.RawMessage `json:"gid"`
+	ID                 json.RawMessage `json:"id"`
+	LocalPasswd        *string         `json:"localPasswd"`
+	PhoneNumber        json.RawMessage `json:"phoneNumber"`
+	PostalCode         json.RawMessage `json:"postalCode"`
+	PublicSSHKey       json.RawMessage `json:"publicSshKey"`
+	Role               json.RawMessage `json:"role"`
+	StateOrProvince    json.RawMessage `json:"stateOrProvince"`
+	TenantID           json.RawMessage `json:"tenantId"`
+	UID                json.RawMessage `json:"uid"`
+	Username           json.RawMessage `json:"username"`
+}
+
+// ValidateAndUnmarshal validates the request and returns a User into which the request's information
+// has been unmarshalled. This allows many fields to be "null", but explicitly checks that they are
+// present in the JSON payload.
+func (u *CurrentUserUpdateRequestUser) ValidateAndUnmarshal(user *User) error {
+	errs := []error{}
+	if u.AddressLine1 != nil {
+		if err := json.Unmarshal(u.AddressLine1, &user.AddressLine1); err != nil {
+			errs = append(errs, fmt.Errorf("addressLine1: %v", err))
+		}
+	}
+
+	if u.AddressLine2 != nil {
+		if err := json.Unmarshal(u.AddressLine2, &user.AddressLine2); err != nil {
+			errs = append(errs, fmt.Errorf("addressLine2: %v", err))
+		}
+	}
+
+	if u.City != nil {
+		if err := json.Unmarshal(u.City, &user.City); err != nil {
+			errs = append(errs, fmt.Errorf("city: %v", err))
+		}
+	}
+
+	if u.Company != nil {
+		if err := json.Unmarshal(u.Company, &user.Company); err != nil {
+			errs = append(errs, fmt.Errorf("company: %v", err))
+		}
+	}
+
+	if u.LocalPasswd != nil && *u.LocalPasswd != "" {
+		if u.ConfirmLocalPasswd == nil || *u.ConfirmLocalPasswd == "" {
+			errs = append(errs, errors.New("confirmLocalPasswd: required when changing password"))
+		} else if *u.LocalPasswd != *u.ConfirmLocalPasswd {
+			errs = append(errs, errors.New("localPasswd and confirmLocalPasswd do not match"))
+		}
+	} else if u.ConfirmLocalPasswd != nil && *u.ConfirmLocalPasswd != "" {
+		errs = append(errs, errors.New("localPasswd: required when changing password"))
+	}
+	user.ConfirmLocalPassword = u.ConfirmLocalPasswd
+	user.LocalPassword = u.LocalPasswd
+
+	if u.Country != nil {
+		if err := json.Unmarshal(u.Country, &user.Country); err != nil {
+			errs = append(errs, fmt.Errorf("country: %v", err))
+		}
+	}
+
+	if u.Email != nil {
+		if err := json.Unmarshal(u.Email, &user.Email); err != nil {
+			errs = append(errs, fmt.Errorf("email: %v", err))
+		} else if err = validation.Validate(*user.Email, is.Email); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if u.FullName != nil {
+		if err := json.Unmarshal(u.FullName, &user.FullName); err != nil {
+			errs = append(errs, fmt.Errorf("fullName: %v", err))
+		} else if user.FullName == nil || *user.FullName == "" {
+			// Perl enforced this
+			errs = append(errs, fmt.Errorf("fullName: cannot be set to 'null' or empty string"))
+		}
+	}
```

Comment:
When a `json.RawMessage` compares like `rm == nil` it means the value was `undefined` - not `null` or anything else "nil-like". If not, it could be `null`, an empty string, an object, an array, NaN, Infinity, -Infinity, or a real number.

Perl treated this endpoint like a PATCH, so I have to distinguish between `null` and `undefined`.

Reply:
Actually, I think NaN and the infinities are extensions of the Python JSON parser, and outside of the JSON spec itself. But that's a bit beside the point, I suppose.

Reply:
Your top PR comment led me to believe that the new Go implementation was intended to more strictly adhere to the semantics of PUT and require a full user representation. Did that intent change or am I misunderstanding something?

Reply:
Oh, yeah, I didn't update the PR. Pursuant to a [mailing list discussion](https://mail-archives.apache.org/mod_mbox/trafficcontrol-dev/201910.mbox/%3CCAMAKGdA5TL6c%3DXJXpB%2BbVj4%3D5gqjUhokqCBFZk20BoUCxYHSxg%40mail.gmail.com%3E) changing this endpoint's behavior - albeit for the better - was shot down, and I re-rewrote it to implement Perl-style PATCH-like behavior.